assertk/src/commonMain/kotlin/assertk/assertions/iterable.kt

```diff
@@ -131,6 +133,34 @@ internal fun MutableList<*>.removeFirst(value: Any?) {
     if (index > -1) removeAt(index)
 }
 
+/**
+ * Asserts the collection contains at least one instance of a given type.
+ *
+ * ```
+ * assertThat(listOf<Any>("one", "two", 1)).containsInstanceOf<String>().each {
+ *   it.hasLength(3)
+ * }
+ * ```
+ */
+inline fun <reified T> Assert<Iterable<*>>.containsInstanceOf(): Assert<List<T>> {
+    return transform("contains subtype of ${T::class}") { actual ->
+        actual.filterIsInstance<T>().also {
```

Comment:
I would use `any` here so that it returns `true` as soon as a single instance is found rather that needing to perform a full iteration and allocate a new list.

Reply:
The sub-list allocation is necessary to provide a way to continue making assertions against the sublist. No?

Reply:
Ah, yeah, I missed the new list was propagated as a return value. I'm a bit hesitant about the API shape using a verb "contains" while the return value is a full filtering.

Reply:
Yeah, agreed. I noticed the `having` discussion after I pushed this. Would `havingInstancesOf` be better? Maybe the negation would be `doesNotHaveInstancesOf`?

Reply:
I guess I'm a bit torn on whether this makes sense as a dedicated operator since it combines a transform with an assertion in a way that I don't think has precedent from other operators. We'll see what the actual maintainers think, though.

Reply:
I'd argue there's precedent with `isNotNull`. It's a transform and an assertion.

Reply:
Fair. I'd find a shorthand for the transformation basically just as helpful.

Reply:
A transformation without assertion might set a (good) precedent for other collection operators.
@evant, could you weigh in when you have some time?